ui/src/DebugReporter.ts

```diff
@@ -10,22 +10,22 @@ export class DebugReporter implements ITelemetryReporter {
     constructor(private _extensionName: string, private _extensionVersion: string, private _verbose: boolean) { }
 
     public sendTelemetryEvent(eventName: string, properties?: { [key: string]: string | undefined; }, measures?: { [key: string]: number | undefined; }): void {
-        this.logTelemetryEvent(eventName, properties, measures, 'TELEMETRY');
+        this.logTelemetryEvent(eventName, properties, measures);
     }
 
-    public sendTelemetryErrorEvent(eventName: string, properties?: { [key: string]: string | undefined; }, measures?: { [key: string]: number | undefined; }): void {
-        this.logTelemetryEvent(eventName, properties, measures, 'TELEMETRY-ERROR');
+    public sendTelemetryErrorEvent(eventName: string, properties?: { [key: string]: string | undefined; }, measures?: { [key: string]: number | undefined; }, _errorProps?: string[]): void {
```

Comment:
Why don't we get rid of this?  We probably want to guard against people accidentally sending error telemetry properties in the standard sendTelemetryEvent anyway, right?

Reply:
You mean get rid of `sendTelemetryEvent` and force people to use `sendTelemetryErrorEvent`? I guess I went with the easiest thing that wouldn't break people, but I suppose it would make sense to remove it if we want to be the enforcers

Reply:
I meant get rid of sendTelemetryErrorEvent, so that sendTelemetryEvent always calls the telemetry's error event.  You only added sendTelemetryErrorEvent in the last PR, right?  So breakage would be minimal?

Seems better than relying on people to call the event right, or else they might violate GDPR...

Reply:
Hmm so you're saying in [here](https://github.com/microsoft/vscode-azuretools/blob/master/ui/src/createTelemetryReporter.ts#L25) we'd have to somehow convert the existing `DebugReporter` we get from `vscode-extension-telemetry`, to a `DebugReporter` that has a method name of `sendTelemetryErrorEvent`, but actually uses `sendTelemetryEvent` under the covers? I'm not sure how I feel about the added layer in the middle. So far, the interface `ITelemetryReporter` directly matches `TelemetryReporter` and only exists so that we can allow a `DebugReporter` with the same type.

That's not what I originally thought you meant, so it doesn't break people how I thought it would. However, it would mean any existing call to `sendTelemetryEvent` would have all properties dropped since `errorProps` is undefined by default

Tbh, I feel like `callWithTelemetryAndErrorHandling` and `registerCommand` does a lot of the "guarding" already. I can't think of a single case where I call `sendTelemetryEvent` directly

Reply:
DebugTelemetry is ours, we don't get it from `vscode-extension-telemetry`.  Instead, we'd need a middle layer like you say to implement ITelemetryReporter to call the TelemetryReporter we get from `vscode-extension-telemetry` [here](https://github.com/microsoft/vscode-azuretools/blob/3167744d999c8e2acd87bd78b8ac895257ebc924/ui/src/createTelemetryReporter.ts#L27).

> So far, the interface ITelemetryReporter directly matches TelemetryReporter and only exists so that we can allow a DebugReporter with the same type.

Well, it also exists to simplify our interface and isolate us from vscode changes.

> I can't think of a single case where I call sendTelemetryEvent directly

Exactly, so why are you *adding* a new method that isn't needed?  We should keep our interface the same (ITelemetryReporter) and have it always do the right behavior, instead of complicating our interface and having it possibly do the wrong thing?

Reply:
Yes I meant to say `TelemetryReporter` for the middle layer

> Exactly, so why are you adding a new method that isn't needed?

For the simplicity of matching `TelemetryReporter`. I agree it's "complicating our interface" in the sense that it has a few more characters, but it does simplify our interface concept-wise in the sense that it has the exact same behavior as the more "official" `TelemetryReporter`

If the goal is to prevent bad behaviors, I'd rather remove `sendTelemetryEvent` from the interface and keep `sendTelemetryErrorEvent`. This should avoid the middle-layer I mentioned and most of the confusion related to differences in `TelemetryReporter` and `ITelemetryReporter`. It would break a _few_ people (as mentioned above I don't think `sendTelemetryEvent` is used very often), but I'd be fine with that if we're doing it to prevent bad behavior. Would that be a satisfactory compromise?

Reply:
Seems weird but I'm okay with it.